File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/globalcontexts/SchemaRuleSet.java
/*
 * Copyright (c) 2015-2022 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */

package com.snowplowanalytics.snowplow.globalcontexts;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snowplowanalytics.snowplow.tracker.InspectableEvent;
import com.snowplowanalytics.snowplow.internal.tracker.SchemaRule;

import java.util.ArrayList;
import java.util.List;

public class SchemaRuleSet {

    @NonNull
    private final List<SchemaRule> rulesAllowed = new ArrayList<>();
    @NonNull
    private final List<SchemaRule> rulesDenied = new ArrayList<>();

    private SchemaRuleSet(@Nullable List<String> allowed, @Nullable List<String> denied) {
        for (String rule : allowed) {
            SchemaRule schemaRule = SchemaRule.build(rule);
            if (schemaRule != null) {
                rulesAllowed.add(schemaRule);
            }
        }
        for (String rule : denied) {
            SchemaRule schemaRule = SchemaRule.build(rule);
            if (schemaRule != null) {
                rulesDenied.add(schemaRule);
            }
        }
    }

    @NonNull
    public static SchemaRuleSet buildRuleSet(@NonNull List<String> allowed, @NonNull List<String> denied) {
        return new SchemaRuleSet(allowed, denied);
    }

    @NonNull
    public static SchemaRuleSet buildRuleSetWithAllowedList(@NonNull List<String> allowed) {
        return buildRuleSet(allowed, new ArrayList<>());
    }

    @NonNull
    public static SchemaRuleSet buildRuleSetWithDeniedList(@NonNull List<String> denied) {
        return buildRuleSet(new ArrayList<>(), denied);
    }

    @NonNull
    public List<String> getAllowed() {
        List<String> result = new ArrayList<>(rulesAllowed.size());
        for (SchemaRule schemaRule : rulesAllowed) {
            result.add(schemaRule.getRule());
        }
        return result;
    }

    @NonNull
    public List<String> getDenied() {
        List<String> result = new ArrayList<>(rulesDenied.size());
        for (SchemaRule schemaRule : rulesDenied) {
            result.add(schemaRule.getRule());
        }
        return result;
    }

    @NonNull
    public FunctionalFilter getFilter() {
        return new FunctionalFilter() {
            @Override
            public boolean apply(@NonNull InspectableEvent event) {
                return matchWithSchema(event.getSchema());
            }
        };
    }

    public boolean matchWithSchema(@Nullable String schema) {
        if (schema == null) {
            return false;
        }
        for (SchemaRule rule : rulesDenied) {
            if (rule.matchWithSchema(schema)) {
                return false;
            }
        }
        if (rulesAllowed.size() == 0) {
            return true;
        }
        for (SchemaRule rule : rulesAllowed) {
            if (rule.matchWithSchema(schema)) {
                return true;
            }
        }
        return false;
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/globalcontexts/SchemaRuleSet.kt
/*
 * Copyright (c) 2015-present Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.globalcontexts

import com.snowplowanalytics.core.tracker.SchemaRule
import com.snowplowanalytics.snowplow.tracker.InspectableEvent

/**
 * Use this class to add a [GlobalContext] to specific events only, based on the event schema.
 * New SchemaRuleSets are created using the companion object `build` methods.
 * 
 */
class SchemaRuleSet private constructor(allowed: List<String?>, denied: List<String?>) {
    private val rulesAllowed: MutableList<SchemaRule> = ArrayList()
    private val rulesDenied: MutableList<SchemaRule> = ArrayList()

    init {
        for (rule in allowed) {
            val schemaRule = rule?.let { SchemaRule.build(it) }
            schemaRule?.let { rulesAllowed.add(schemaRule) }
        }
        for (rule in denied) {
            val schemaRule = rule?.let { SchemaRule.build(it) }
            schemaRule?.let { rulesDenied.add(schemaRule) }
        }
    }

    val allowed: List<String>
        get() {
            val result: MutableList<String> = ArrayList(rulesAllowed.size)
            for (schemaRule in rulesAllowed) {
                result.add(schemaRule.rule)
            }
            return result
        }
    
    val denied: List<String>
        get() {
            val result: MutableList<String> = ArrayList(rulesDenied.size)
            for (schemaRule in rulesDenied) {
                result.add(schemaRule.rule)
            }
            return result
        }
    
    val filter: FunctionalFilter
        get() = object : FunctionalFilter() {
            override fun apply(event: InspectableEvent): Boolean {
                return matchWithSchema(event.schema)
            }
        }

    fun matchWithSchema(schema: String?): Boolean {
        if (schema == null) return false
        
        for (rule in rulesDenied) {
            if (rule.matchWithSchema(schema)) return false
        }
        
        if (rulesAllowed.size == 0) return true
        
        for (rule in rulesAllowed) {
            if (rule.matchWithSchema(schema)) return true
        }
        return false
    }

    /**
     * Build a [SchemaRuleSet]. Provide schema strings for events that entities will either be attached to
     * (`allowed` schemas) or not (`denied` schemas).
     * Schema strings can contain wildcards, e.g. 
     * `iglu:com.acme.marketing / * / jsonschema / *-*-*` (without spaces).
     * 
     * They follow the same five-part format as an Iglu URI `iglu:vendor/event_name/format/version`,
     * with the exception that a wildcard can be used to refer to all cases.
     * 
     * The parts of a rule are wildcarded with certain guidelines:
     * - asterisks cannot be used for the protocol (i.e. schemas always start with `iglu:`)
     * - version matching must be specified like so: `––`, where any part of the versioning can be defined, e.g. `1-–`, 
     * but only sequential parts can be wildcarded, e.g. `1--1` is invalid but `1-1–*` is valid
     * - vendors cannot be defined with non-wildcarded parts between wildcarded parts: `com.acme.*.marketing.*` is invalid, while `com.acme.*.*` is valid
     */
    companion object {
        @JvmStatic
        fun buildRuleSet(allowed: List<String>, denied: List<String>): SchemaRuleSet {
            return SchemaRuleSet(allowed, denied)
        }

        @JvmStatic
        fun buildRuleSetWithAllowedList(allowed: List<String>): SchemaRuleSet {
            return buildRuleSet(allowed, ArrayList())
        }

        @JvmStatic
        fun buildRuleSetWithDeniedList(denied: List<String>): SchemaRuleSet {
            return buildRuleSet(ArrayList(), denied)
        }
    }
}
